### PR TITLE
feat(protocol): add ProverSet to hold TKO tokens

### PR DESCRIPTION
**Pull request by dantaik**
Why I added this contract? I hope our own provers (EOAs) don't hold TKO token directly. This ProverSet contract can now hold all TKO tokens to be used as bonds.

@davidtaikocha we should test this contract in Hekla first.

### CODE REVIEW

File: packages/protocol/contracts/team/provers/MyProver.sol [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+        emit ProverEnabled(_prover, _isProver);
+    }
+
+    /// @notice Withdraws Taiko tokens back to the owner address.
+    function withdraw(uint256 _amount) external onlyOwner {
+        IERC20(TKO).transfer(owner(), _amount);
+    }
+
+    /// @notice Proves a Taiko block.
```
Comment (dantaik) [inline packages/protocol/contracts/team/provers/MyProver.sol:45]:
```suggestion
    /// @notice Proves or contests a Taiko block.
```

**Review by Brechtpd — APPROVED** [review]

File: packages/protocol/contracts/team/proving/ProverSet.sol [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "../../common/EssentialContract.sol";
+import "../../L1/ITaikoL1.sol";
+
+/// @title ProverSet
+/// @notice A contract that holds TKO token and acts as a Taiko prover. This contract will simply
+/// relay `proveBlock` calls to TaikoL1 so msg.sender doesn't need to hold any TKO.
+/// @custom:security-contact security@taiko.xyz
+contract ProverSet is EssentialContract, IERC1271 {
```
Comment (adaki2004) [inline packages/protocol/contracts/team/proving/ProverSet.sol:14]: I think in order to transfer some TKO on behalf of `this` contract to `TaikoL1`  via `AssignmentHook` we shall have a function something like:

    function approveTaikoL1Contract(
        uint256 amount
    )
        public
        onlyOwner
    {
        IERC20(tko).approve(TAIKO, amount);
    }

File: packages/protocol/contracts/team/proving/ProverSet.sol [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    bytes4 private constant _EIP1271_MAGICVALUE = 0x1626ba7e;
+    address public constant TKO = 0x10dea67478c5F8C5E2D90e5E9B26dBe60c54d800;
+    address public constant TAIKO = 0x06a9Ab27c7e2255df1815E6CC0168d7755Feb19a;
+    address public constant ASSIGNMENT_HOOK = 0x537a2f0D3a5879b41BCb5A2afE2EA5c4961796F6;
```
Comment (adaki2004) [inline packages/protocol/contracts/team/proving/ProverSet.sol:18]: Maybe better do the resolve via the address manager ? (If we want to test out on Hekla too)

Comment (dantaik) [reply packages/protocol/contracts/team/proving/ProverSet.sol:18]: good suggestion.  Will make the change.

File: packages/protocol/contracts/team/proving/ProverSet.sol (hunk 2 above)
Comment (dantaik) [reply packages/protocol/contracts/team/proving/ProverSet.sol:15]: The init function set allowance for TaikoL1 and the AssignmentHook to the max.

File: packages/protocol/contracts/team/proving/ProverSet.sol (hunk 3 above)
Comment (dantaik) [reply packages/protocol/contracts/team/proving/ProverSet.sol]: Done

File: packages/protocol/deployments/mainnet-contract-logs.md [hunk 4]
```diff
@@ -155,6 +155,8 @@
   - admin.taiko.eth accepted the ownership @tx`0x0ed114fee6de4e3e2206cea44e6632ec0c4588f73648d98d8df5dc0183b07885`
   - Upgraded from `0xd912aB787624c9eb96a37e658e9596e114360440` to `0xF1cA1F1A068468E1dcF90dA6add185467de80943` @commit `b90b932` @tx`0x416560cd96dc75ccffebe889e8d1ab3e08b33f814dc4a2bf7c6f9555071d1f6f`
   - `Init2()` called @tx `0x7311fee56f87294e336393b55939489bc1e810c402f304013475d04c90ca32a9`
+- todo:
```
Comment (dantaik) [inline packages/protocol/deployments/mainnet-contract-logs.md:158]: @davidtaikocha

File: packages/protocol/contracts/team/proving/ProverSet.sol (hunk 2 above)
Comment (adaki2004) [reply packages/protocol/contracts/team/proving/ProverSet.sol:15]: Ah yeah, i see now!

**Review by adaki2004 — APPROVED** [review]

**Pull request merged**